designs/odcr.md

```diff
@@ -0,0 +1,121 @@
+
+Each Capacity Reservation is defined with:
+
+- The Availability Zone in which to reserve the capacity
+- The number of instances for which to reserve capacity
+- The instance attributes, including the instance type, tenancy, and platform/OS
+- Instance match criteria
+  - If Capacity Reservation is targeted, only accept instances that matches all attributes + explicitly targeted the capacity reservation
+  - Open if capacity reservation accepts all instances that matches all attributres
+
+AWS also supports grouping Capacity Reservation into Capacity Reservation groups. 
+Both these entities are supported in Launch Template's CapacityReservationTarget [definitions](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ec2-launchtemplate-capacityreservationtarget.html).
+
+## Goals
+
+- Support associating ODCR to EC2NodeClass
+- Define Karpenter's behavior when launching nodes into Capacity Reservation
+- Define Karpenter's behavior when encountering errors when attempting to launch nodes into Capacity Reservation
+
+## Non-Goals
+
+_We are keeping the scope of this design very targeted so even if these could be things we eventually support, we aren't scoping them into this design_
+- Supporting prioritization when launching nodes 
+- Supporting any pre-calcaluation of ODCR capacity utilization before node launch
+- Supporting changes in scaling behavior when ODCR is associated to a NodeClass. _We won't bring up N nodes to match an N node capacity reservation_
+
+## Proposed Solution
+
+### Supporting associating ODCR to EC2NodeClass
+
+Add a new field `capacityReservationSpec` to `EC2NodeClass` 
+```yaml
+apiVersion: karpenter.k8s.aws/v1beta1
+kind: EC2NodeClass
+metadata:
+  name: example-node-class
+spec:
+  capacityReservationSpec:
+    capacityReservationPreference: open | none | None  # Cannot be defined if capacityReservationTarget is specified
+    capacityReservationTarget: # Cannot be defined if capacityReservationPreference is specified
+      capacityReservationId: String | None
+      capacityReservationResourceGroupArn: String | None
+```
+This follows exactly how LaunchTemplate defines [CapacityReservationSpecification](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ec2-launchtemplate-capacityreservationspecification.html) with the same
+constraints. 
+
+Karpenter will perform validation against the spec to ensure there isn't any violation prior to creating the LaunchTemplates.
+
+### Launching Nodes into Capacity Reservation
+
+#### All Launch Templates are associated with the specified Capacity Reservation
+
+EC2NodeClass currently supports automatically generating Launch Templates based on instance types and their AMI requirements however as the first iteration of supporting Capacity Reservation, we won't introduce any prioritization nor fallback nor safeguard when launching nodes against EC2NodeClasses with a Capacity Reservation. This means that it is the
```

Comment:
I personally think that we should consider the fallback to be cross-NodePool rather than within the NodePool itself. If Karpenter is aware of how many instances are in a capacity reservation and it also knows that we are at the cap, then it can know to not try to launch this NodePool because no instance types are available anymore and it would fallback to the next lowest-weighted NodePool. That would be similar to our handling for spot preference with on-demand fallback today

Reply:
I am definitely not super familiar with the fallback logic so I will read up on it. My thought was calling then failing to launch due to insufficient capacity has similar functional and performance to pre-checking then falling back to other node pools

Reply:
> I personally think that we should consider the fallback to be cross-NodePool rather than within the NodePool itself

I'll clarify what I mean on this here since I think there has been some updated thinking since this RFC was introduced. IMO, we should have the following:
1. A user should be able to fallback between ODCR -> Spot -> OD in the same NodePool and can define a superset of instance types beyond the ones that have ODCRs, even if they are specifying capacityReservationGroupSelectorTerms
2. Multiple Capacity Reservations should be supported within a single EC2NodeClass
3. Any capacity reservations that are discovered through the capacityReservationGroupSelectorTerms should be added to the set of available offerings that are returned by the GetInstanceTypes() call. This should just _add_ additional offerings (with a price reflecting the fact that the capacity is already paid for -- maybe 0) on top of the already surfaced offerings for spot and standard OD